### ISSUE TITLE
Events Published to virtual method with SubscriberAttribute not calling overriden methods

### ISSUE THREAD

**Issue by captkirk88** [Doraku/DefaultEcs#38]
Reproduce by creating a Dummy with a virtual method.  Derive from Dummy with a new FlailingDummy, Subscriber method not called by world.Publish.

The reflection strategy doesn't appear to detect SubscriberAttributes on overriden virtual methods.

**Comment by Doraku** [comment]
Can you give the code of your classes? I tried both way (base class method decorated and derived class method decorated) and it seems to behave correctly. Maybe my test is wrong?

**Comment by captkirk88** [comment]
Pretty much goes...
```csharp
EntitySystem (abstract class) => public virtual [Subscribe]OnAdded(in EntityAdded msg) => 
MovementSystem (inherits EntitySystem) => public override OnAdded(...) 
```
I have a InitializeAll that initializes and subscribes each EntitySystem, via World#Subscribe which should register the attributes.
```csharp
                if (updateSystems[i] is EntitySystem sys)
                {
                    sys.Initialize();
                    World.Subscribe(sys);
                    Logger.Debug<EntityManager>($"{sys.GetType().Name} initialized");
                }
```
Then when a Entity is created and all the components are added, I call EntityManager#Commit and it sends the `World#Publish(new EntityAdded(in ent));`.

**Comment by captkirk88** [comment]
Also the EntitySystem and MovementSystem are in separate dll's.

**Comment by Doraku** [comment]
It's strange, this scenario should definitely work and I don't see how separate dlls or the implicit conversion could impact anything since you are publishing explicitly an EntityAdded message. Obviously I suppose the World instance you are using to subscribe and to publish are the same, I'll try to see if I can replicate.

**Comment by captkirk88** [comment]
Do you think threads would have an effect on it's ability to sync up somewhere around 65,000 messages?  A reasonable number for a scene of particles, for example.

**Comment by Doraku** [comment]
I found a bug there (GetCustomAttribute should not be passed true or you'll have multiple subscriptions on overriden method if the base one is decorated, stupid me >_>)
https://github.com/Doraku/DefaultEcs/blob/6ab9238c28374c867099c65d3cc064549d6bebda/source/DefaultEcs/IPublisherExtension.cs#L38
but I don't think it is linked to your current problem, since the method isn't even called once.

The number of messages shouldn't change anything. The publisher is used internally to notify EntitySet of Entity modifications so everything should have broke much earlier if this was the case.

What are you doing in parallel? Most methods on Entity are not thread safe (and World.CreateEntity isn't either), only Is..., Has<>, Get<> and Set<> (once a component has already been set) are thread safe.

**Comment by captkirk88** [comment]
Entities themselves are being called in a threadsafe context.  It is the World#Publish call that is being called in a separate thread from the main thread.  Calling the EntitySystem#OnAdded directly works as intended.  

I wonder if with the true flag it is removing the subscriptionattribute since it is not assigned to the overridden virtual method.

Also, I would use ReflectionOnly* methods where possible.  The difference being in performance with reflection lookup.

**Comment by captkirk88** [comment]
Huh, that was the case.
```csharp
        [Test, Category("Github")]
        public void VirtualMethodLookup()
        {
            var mi = typeof(Derived).GetMethod("TestMethod");
            Assert.True(mi.GetCustomAttribute<Shared.Core.InterceptorAttribute>(false) != null);
            Assert.True(mi.GetCustomAttribute<Shared.Core.InterceptorAttribute>(true) != null);
        }

        private class Base
        {
            [Shared.Core.Interceptor]
            public virtual void TestMethod()
            {
                Console.WriteLine("WOO");
            }
        }

        private class Derived : Base
        {
            public override void TestMethod()
            {
                Console.WriteLine("REEEE");
            }
        }
```

**Comment by captkirk88** [comment]
Having that set to true actually works for my scenario.  It should anyways.  I tested myself with a unittest to verify having it set to true resulted in the base type being checked, it does.  The question is what are you doing so differently from this test.
```csharp
        [Test, Category("Github")]
        public void VirtualMethodLookup()
        {
            var mi = typeof(Derived).GetMethod("TestMethod");
            Assert.True(mi.GetCustomAttribute<Shared.Core.InterceptorAttribute>(true) != null);
            Assert.False(mi.GetCustomAttribute<Shared.Core.InterceptorAttribute>(false) != null);
        }

        private class Base
        {
            [Shared.Core.Interceptor]
            public virtual void TestMethod()
            {
                Console.WriteLine("WOO");
            }
        }

        private class Derived : Base
        {
            public override void TestMethod()
            {
                Console.WriteLine("REEEE");
            }
        }
```

**Comment by captkirk88** [comment]
Yours passed... this is very puzzling.  It might be one of those other checks you do in that method.
```csharp
        [Test, Category("Github")]
        public void WhatYouAreDoing()
        {
            var target = new Derived();
            Assert.NotZero(typeof(Derived).GetTypeInfo().DeclaredMethods
                .Count(m => m.GetCustomAttribute<Shared.Core.InterceptorAttribute>(true) != null && (m.IsStatic || target != null)));
        }

```

**Comment by captkirk88** [comment]
Yet this fails...
```csharp
            var target = new Derived();
            int goodCount = 0;
            foreach (MethodInfo method in target.GetType().GetTypeInfo().DeclaredMethods
                .Where(m => m.GetCustomAttribute<SubscribeAttribute>(true) != null && (m.IsStatic || target != null)))
            {
                ParameterInfo[] parameters = method.GetParameters();

                if (parameters.Length != 1
                    || !parameters[0].ParameterType.IsByRef
                    || method.ReturnType != typeof(void))
                {
                    throw new NotSupportedException($"Can't apply {nameof(SubscribeAttribute)} to \"{method.Name}\"");
                }

                Type argType = parameters[0].ParameterType.GetElementType();
                goodCount++;
            }
            Assert.NotZero(goodCount);
```

**Comment by captkirk88** [comment]
Is it possible that this is not performing as intended with virtual methods?
```csharp
                        subscriptions.Add((IDisposable)_subscribeMethod.MakeGenericMethod(argType).Invoke(
                            publisher,
                            new object[]
                            {
                            method.IsStatic
                                ? method.CreateDelegate(typeof(ActionIn<>).MakeGenericType(argType))
                                : method.CreateDelegate(typeof(ActionIn<>).MakeGenericType(argType), target)
}));
```

**Comment by Doraku** [comment]
Sorry I was not much available this week end.

> Yours passed... this is very puzzling. It might be one of those other checks you do in that method.

This is because I look up base type also, that's why I don't need to inherit attributes in the lookup so I don't perform multiple subscribe for the same method
```
                while (type != null)
                {
                    ...

                    type = type.GetTypeInfo().BaseType;
                }
```

If this is not too much hassle can you get the latest source, reference them directly in your project and debug the subscribe call to see exactly what is happening? I'll try to make a more real test on my side.

**Comment by captkirk88** [comment]
Seems that VS hates me today.  Getting the upgrade tool popup for that project format.  Tried even adding it via dotnet sln add <PATH> and no go.  Using VS 2017.  But that would be a separate issue.  I am going to run some more thorough analysis in trying to figure out the issue as well, just without the source project to debug into.  Fun times.

**Comment by Doraku** [comment]
```
        public abstract class AbstractDecorated
        {
            [Subscribe]
            protected virtual void Do(in int i) => Console.WriteLine($"abstract {i}");
        }

        public class Inherit : AbstractDecorated
        {
            protected override void Do(in int i) => Console.WriteLine($"derived {i}");
        }

        private static void Main()
        {
            using (IPublisher world = new World())
            {
                object o = new Inherit();

                world.Subscribe(o);

                world.Publish(42);
            }
        }
```

This works fine and output `derived 42`. The only things I could think of is that either your are not publishing on the same World instance as the one used to subscribe, or you have duplicate EntityAdded types in different namespace and you are not publishing the one you expect :/

**Comment by captkirk88** [comment]
Closing this, turns out it was on my end.  Sorry for the waste of an issue.

**Comment by Doraku** [comment]
No worries there was indeed a bug in there so in wasn't a waste, glad you sorted your problem out in the end ^^
